saleor/checkout/calculations.py

```diff
@@ -496,6 +510,55 @@ def _fetch_checkout_prices_if_expired(
     return checkout_info, lines
 
 
+@allow_writer()
+def recalculate_discounts(
+    checkout_info: "CheckoutInfo",
+    lines_info: Iterable["CheckoutLineInfo"],
+    database_connection_name: str = settings.DATABASE_CONNECTION_DEFAULT_NAME,
+    force_update: bool = False,
+) -> tuple["CheckoutInfo", Iterable["CheckoutLineInfo"]]:
+    """Recalculate checkout discounts.
+
+    Discounts are recalculated only if force_update is True, or if both discount
+    and price expirations have passed.
+    This updates catalogue promotions, vouchers, and order promotion discounts.
+    """
+    checkout = checkout_info.checkout
+
+    # Do not recalculate discounts in case the checkout prices are still valid, either
+    # discounts or tax prices.
+    if not force_update and (
+        checkout.discount_price_expiration > timezone.now()
+        or checkout.price_expiration > timezone.now()
```

Comment:
Extra check that prevents running discount recalculation in case the checkout prices don't expire.

Reply:
Hhm, I think that we need something like one screenshot (just to show the intention): 
<img width="672" height="350" alt="image" src="https://github.com/user-attachments/assets/2ed1ea98-d2e6-40e9-b512-5acee209fa0b" />

AFAIK we should update the discount, when they are expired. If something has been changed then we should mark prices as invalid.